pkg/controller/core/leader_aware_reconciler.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	config "sigs.k8s.io/kueue/apis/config/v1beta1"
+)
+
+// WithLeadingManager returns a decorating reconcile.Reconciler that discards reconciliation requests
+// for the controllers that are started with the controller.Options.NeedLeaderElection
+// option set to false in non-leading replicas.
+//
+// Starting controllers in non-leading replicas is needed for these that update the data
+// served by the visibility extension API server.
+//
+// This enables to:
+//   - Keep the scheduling decisions under the responsibility of the leading replica alone,
+//     to prevent any concurrency issues.
+//   - Consume requests from the watch event queues, to prevent them from growing indefinitely
+//     in the non-leading replicas.
+//   - Transition to actually reconciling requests in the replica that may acquire
+//     the leader election lease, in case the previously leading replica failed to renew it.
+func WithLeadingManager(mgr ctrl.Manager, reconciler reconcile.Reconciler, cfg *config.Configuration) reconcile.Reconciler {
+	// Do not decorate the reconciler if leader election is disabled
+	if cfg.LeaderElection == nil || !ptr.Deref(cfg.LeaderElection.LeaderElect, false) {
+		return reconciler
+	}
+
+	// Default to the recommended lease duration, that's used for core components
+	requeueDuration := 15 * time.Second
+	// Otherwise used the configured lease duration for the manager
+	if le := cfg.LeaderElection; le != nil {
+		zero := metav1.Duration{}
+		if duration := le.LeaseDuration; duration != zero {
+			requeueDuration = duration.Duration
+		}
```

Comment:
Is this check and defaults needed?

We already check this duration here: https://github.com/kubernetes-sigs/kueue/blob/4bc332a75fe221e6dc92db269ae3832da884527a/pkg/config/config.go#L119-L121

Also, the controller-runtime defaults to 15sec if this duration is zero here:

https://github.com/kubernetes-sigs/controller-runtime/blob/8f8247fdc1c3bc75f2189f28079a9efc4e82d06f/pkg/manager/manager.go#L204-L206

Reply:
I haven't been able to find where we default the value for `cfg.LeaderElection.LeaseDuration.Duration`.

Maybe we should call `RecommendedDefaultLeaderElectionConfiguration` from component-base when we initialise the configuration, but I'd rather to that in a separate PR, WDYT?

Reply:
> I haven't been able to find where we default the value for cfg.LeaderElection.LeaseDuration.Duration.

We default the leaderElection parameters here:

https://github.com/kubernetes-sigs/kueue/blob/727e0b522e0bd3e1a785be13430fcf29a34b3393/apis/config/v1beta1/defaults.go#L76

> Maybe we should call RecommendedDefaultLeaderElectionConfiguration from component-base when we initialise the configuration, but I'd rather to that in a separate PR, WDYT?

You mean the below helper, right? This should work fine. I'm ok with a separate PR.

https://github.com/kubernetes/component-base/blob/d5ddea0e47af10468dcd90e09768b2661caf23f7/config/v1alpha1/defaults.go#L35

Reply:
> We default the leaderElection parameters here

I've seen that, but I fail to see how this would prevent `cfg.LeaderElection` from being `nil`. Am I missing something?

> You mean the below helper, right? This should work fine. I'm ok with a separate PR.

Right. I'll open a separate PR ASAP then.

Reply:
> You mean the below helper, right? This should work fine. I'm ok with a separate PR.

I've created #1598. I can rebase this one on top once it's merged.

Reply:
Addressed in #1652.